test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs

```diff
@@ -1024,6 +1024,180 @@ public void ReferenceCountingTest_MultiThreaded()
             TestUtilities.AssertFailIfErrors(context);
         }
 
+        /// <summary>
+        /// Testing adding/removing providers to the Default cache w/o leaking task at the end of test.
+        /// </summary>
+        [Fact]
+        public void ProviderCacheTest_EnsureNoLeakingTasks()
+        {
+            int waitTimeoutInSeconds = 60;
+
+            var cache = new InMemoryCryptoProviderCache();
+            var factory = new CryptoProviderFactory(cache);
+
+            // create signing providers
+            var signingProviders = CreateSigningProviders(factory);
+
+            // create verifying providers
+            var verifyingProviders = CreateVerifyingProviders(factory);
+
+            // make sure providers can be retrieved from the cache
+            if (cache.TryGetSignatureProvider(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm, typeof(AsymmetricSignatureProvider).ToString(), true, out var tmpProvider))
+            {
+                Assert.True(tmpProvider != null);
+            }
+
+            Thread.Sleep(3000); // give it some time to allow the tasks to start
+
+            Assert.True(cache.TaskCount > 0, $"ProviderCacheTest_EnsureNoLeakingTasks: unexpected task count: {cache.TaskCount}, expected: > 0");
+
+            // remove all signing providers
+            foreach (var provider in signingProviders)
+                cache.TryRemove(provider);
+
+            foreach (var provider in verifyingProviders)
+                cache.TryRemove(provider);
+
+            WaitTillTasksComplete(cache, waitTimeoutInSeconds); // wait up to 60 seconds for the task(s) to complete
+
+            Assert.True(cache.TaskCount == 0, $"ProviderCacheTest_EnsureNoLeakingTasks: unexpected task count: {cache.TaskCount}, expected: 0");
```

Comment:
Wondering whether this will be sufficient to fully address #1650.
Would it be required for users to manually cleanup the cache in each test?

Reply:
This will not address the issue in #1650. I think we need to document that users need to clean up the cache in order for the task to complete. Do you know where to describe the behaviors other than in #1650?